src/field/gen_fieldops.py

```diff
@@ -0,0 +1,338 @@
+    for dimension, index in dims.items():
+        for_loop += for_loop_statement_template.format(dimension=dimension,
+                                                       index=index)
+
+    # Check that the mesh is the same for all fields
+    mesh_equality_assert = ""
+    compound_mesh_equality_assert = ""
+    if lhs != 'BoutReal' and rhs != 'BoutReal':
+        mesh_equality_assert = "ASSERT1(localmesh == rhs.getMesh());"
+        compound_mesh_equality_assert = "ASSERT1(fieldmesh == rhs.getMesh());"
+
+    # Either total number of elements, or size of each dimension separately
+    m = ',' if elementwise else '*'
+    non_compound_length_arg = m.join(["localmesh->LocalN{}".format(d) for d in out.dimensions])
+    compound_length_arg = m.join(["fieldmesh->LocalN{}".format(d) for d in out.dimensions])
+
+    # Only check field location for Field3D
+    non_compound_location_check = ""
+    compound_location_check = ""
+    if lhs == rhs == 'Field3D':
+        non_compound_location_check = location_check_template.format(operator_name=operator_name)
+        compound_location_check = compound_location_check_template.format(
+            operator_name=operator_name)
+
+    # Set location for result (again, only for Field3d)
+    non_compound_location_set = ""
+    if out == 'Field3D':
+        source = "rhs" if rhs == "Field3D" else "lhs"
+        non_compound_location_set = location_set_template.format(location_source=source)
+
+    mesh_source = "lhs" if lhs != 'BoutReal' else "rhs"
+
+    template_args = {
+        'out': out,
+        'lhs': lhs,
+        'rhs': rhs,
+
+        'operator': operator,
+        'operator_name': operator_name,
+
+        'low_level_length_arg': low_level_length_arg,
+        'non_compound_length_arg': non_compound_length_arg,
+        'compound_length_arg': compound_length_arg,
+
+        'for_loop': for_loop,
+
+        'mesh_equality_assert': mesh_equality_assert,
+        'compound_mesh_equality_assert': compound_mesh_equality_assert,
+
+        'mesh_source': mesh_source,
+        'non_compound_location_check': non_compound_location_check,
+        'non_compound_location_set': non_compound_location_set,
+        'compound_location_check': compound_location_check,
+
+        'result_op': out.getElement(data=elementwise),
+        'lhs_op': lhs.getElement(data=elementwise),
+        'rhs_op': rhs.getElement(data=elementwise),
+    }
+
+    return template_args
+
+
+if __name__ == "__main__":
+    for_gcc = True
+    print(header)
+
+    for lhs, rhs in itertools.product(fields, fields):
+        # We don't have define real real operations
+        if lhs == rhs == 'BoutReal':
+            continue
+        rhs = copy(rhs)
+        lhs = copy(lhs)
+
+        # The output of the operation. The `larger` of the two fields.
+        out = returnType(rhs, lhs)
+        out.name = 'result'
+        lhs.name = 'lhs'
+        rhs.name = 'rhs'
+
+        for operator, operator_name in operators.items():
```

Comment:
`operators` is a dict, and does not guarantee that the order remains the same for different runs. Thus the generated code is not reproducible.

Reply:
Good point, I've been using python 3.6 where this was an implementation detail. I've switched to using `OrderedDict` so this will always work

Reply:
Thanks for fixing my bug 👍